### PR TITLE
feat: local download dir for Model and Estimator classes

### PR DESCRIPTION
**Pull request by evakravi**
*Issue #, if available:*

*Description of changes:*
Adds `local_download_dir` argument to `Estimator` and `Model` classes. This allows users the ability to specify a directory for downloading temporary artifacts. This change was motivated by the need to use a non-default download location when using the SageMaker Python SDK in SageMaker Studio (disk space limit errors were experienced).

*Testing done:*
Added unit tests. 

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/fw_utils.py [hunk 1]
```diff
@@ -418,7 +421,13 @@ def tar_and_upload_dir(
     script_name = script if directory else os.path.basename(script)
     dependencies = dependencies or []
     key = "%s/sourcedir.tar.gz" % s3_key_prefix
-    tmp = tempfile.mkdtemp()
+    if local_download_dir is not None and not (
+        os.path.exists(local_download_dir) and os.path.isdir(local_download_dir)
+    ):
+        raise ValueError(
+            f"Inputted directory for storing newly generated temporary directory does not exist: '{local_download_dir}'"
+        )
```
Comment (JGuinegagne) [inline src/sagemaker/fw_utils.py:429]: how about attempting to create the target dir with `os.makedirs(dir, exists_ok=True)` instead?

File: src/sagemaker/utils.py [hunk 2]
```diff
@@ -368,11 +368,17 @@ def _tmpdir(suffix="", prefix="tmp"):
...
 
     Returns:
         str: path to the directory
     """
-    tmp = tempfile.mkdtemp(suffix=suffix, prefix=prefix, dir=None)
+    if dir is not None and not (os.path.exists(dir) and os.path.isdir(dir)):
+        raise ValueError(
+            f"Inputted directory for storing newly generated temporary directory does not exist: '{dir}'"
+        )
+    tmp = tempfile.mkdtemp(suffix=suffix, prefix=prefix, dir=dir)
```
Comment (JGuinegagne) [inline src/sagemaker/utils.py:381]: same comment here

**Review by JGuinegagne — CHANGES_REQUESTED** [review]
As discussed, `Model` and `Estimator` already have too many arguments IMO.

I would vote for using [session_settings](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/session_settings.py) instead to behave as an optional override to the default behavior (which is to write to the `/tmp` dir). Example usage [here](https://github.com/aws/sagemaker-python-sdk/blob/438984754a8f44b34d70154197a3bbeb0272f052/src/sagemaker/fw_utils.py#L422).

Could you please also explain in the PR how you tested this? Ideally:
- package the SDK, `pip install` in `jumpstart-env` conda env in Studio instance, then restart the JupyterServer and verify.
- update the JumpStart server to override this setting and verify it works as expected

- `pip install` SDK tarball in a notebook and verify that the repack still work as expected (no override).

File: src/sagemaker/fw_utils.py (hunk 1 above)
Comment (evakravi) [reply src/sagemaker/fw_utils.py:429]: I'd rather not. Making directories on behalf of customers seems risky.

**Review by JGuinegagne — APPROVED** [review]
LGTM, but could you please clarify why `SessionSettings()` is necessary for all these unit tests now whereas it wasn't the case before?

**Review by trajanikant — COMMENTED** [review]
/bot run all

**Review by trajanikant — COMMENTED** [review]
/bot run pr, slow-tests

**Review by trajanikant — COMMENTED** [review]
/bot run slow-tests

**Review by trajanikant — COMMENTED** [review]
/bot run slow-tests

**Review by trajanikant — COMMENTED** [review]
/bot run slow-tests

**Review by trajanikant — COMMENTED** [review]
/bot run notebook-tests, slow-tests

**Review by trajanikant — COMMENTED** [review]
/bot run notebook-tests, slow-tests

**Pull request closed without merging**